c/driver/postgresql/connection.cc

```diff
@@ -182,7 +216,12 @@ AdbcStatusCode PostgresConnectionGetObjectsImpl(
         if (depth == ADBC_OBJECT_DEPTH_CATALOGS) {
           CHECK_NA(INTERNAL, ArrowArrayAppendNull(catalog_db_schemas_col, 1), error);
         } else {
-          return ADBC_STATUS_NOT_IMPLEMENTED;
+          if (depth >= ADBC_OBJECT_DEPTH_DB_SCHEMAS) {
+            RAISE_ADBC(PostgresConnectionGetSchemasImpl(conn, depth,
```

Comment:
Here we are grabbing the associated schemas in a loop, but this brings up a structural question. I assume from the adbc.h header that GetInfo should maintain a hierarchy from catalog -> schema -> table -> constraints, etc...

However, the current implementation (shared with sqlite) creates the overall array structure in a call to `AdbcInitConnectionObjectsSchema`, but at face value I don't see how this preserves the hierarchy between elements. With this happening in a loop, it ends up writing the schema name for as many databases as exist. I could certainly move it out of the loop to fix that, but figured out ask about the hierarchical intent of the result before moving too much around

Reply:
Yeah, it's supposed to be a nested data structure, so the catalog name appears once, with the schemas nested inside that row, and ditto for tables being nested into catalogs. The Flight SQL driver has code along those lines, if you want to reference it (in Go/Java).

It is a bit annoying and if I could do it again I would probably stick with the Flight SQL design (GetTables/GetCatalogs/GetSchemas). (Or possibly Calcite's design, which is more flexible; that has come up before too.)

Reply:
AFACIT those implementations generate maps that relate a table to a schema. I'm guessing elsewhere in the FlightSQL code base is where that mapping gets generated back into an Arrow structure. Any pointers on where I should look for that?

Generally I am unsure I have the right idea of how this is all supposed to be laid out. For example, if we have the following setup in the database:

```json
[{
 "catalog_name": "postgres",
 "catalog_db_schema": [
 {
   "db_schema_name": "public",
   "db_schema_tables": [...],
 },
 {
   "db_schema_name": "template0",
   "db_schema_tables": [...],
 },
 {
   "db_schema_name": "template1",
   "db_schema_tables": [...],
 },
 }]
 }]
```

Am I supposed to access the value of "public" via `reader->array_view->children[1][0]->children[0]`? I have a feeling the `children[1][0]` access is incorrect and there should be a way to extract the list element by index from an array via nanoarrow, but didn't see such a function. 

Is that something that needs to be implemented upstream or do I have the wrong mental model?

Reply:
So to iterate "catalog_db_schema", you would use nanoarrow to get the array offsets of element 0 in the catalog_db_schema array, which would give you a slice of the child array to iterate through. For each of those elements, you would then again get the array offsets of that element in the db_schema_tables array, which would give you a slice of the child child array.

Reply:
You can see this starting here: https://github.com/apache/arrow-adbc/blob/47775a82caf49b9c7a706717f4204cd98b5efcf8/c/validation/adbc_validation.cc#L564

The code I'm referencing for building up the structure is here: https://github.com/apache/arrow-adbc/blob/main/go/adbc/driver/internal/shared_utils.go

You could also reference the code here: https://github.com/apache/arrow/pull/14082

Reply:
> So to iterate "catalog_db_schema", you would use nanoarrow to get the array offsets of element 0 in the catalog_db_schema array, which would give you a slice of the child array to iterate through...

Great thanks for the guidance. I see nanoarrow has `ArrowArrayViewUnionChildOffset` but didn't see anything similar for Lists (though I think the binary layout would be the same as far as lists go). Think its worth trying to upstream that function for lists first?

Reply:
Ah...I was going to say it exists, but just realized, it's actually part of this project: https://github.com/apache/arrow-adbc/blob/47775a82caf49b9c7a706717f4204cd98b5efcf8/c/validation/adbc_validation_util.cc#L67

I think it'd be worth upstreaming.

Reply:
Cool - I'll give that a shot